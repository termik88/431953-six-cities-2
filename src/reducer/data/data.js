import {preparePlacesData, preparePlace, getCitiesList, getRandomCity} from "../../until.js";

const REQUEST_URL = {
  HOTELS: `/hotels`,
  FAVORITE: `/favorite`
};

const initialState = {
  cityCurrent: ``,
  citiesList: [],
  placesAll: [],
};

const ActionType = {
  CHANGE_CITY_CURRENT: `CHANGE_CITY_CURRENT`,
  LOAD_PLACES_DATA: `LOAD_PLACES_DATA`,
  ADD_FAVORITE_PLACE: `ADD_FAVORITE_PLACE`
};

const ActionsCreator = {
  changeCityCurrent: (cityName) => ({
    type: ActionType.CHANGE_CITY_CURRENT,
    payload: cityName,
  }),

  loadData: (placesAll, citiesList, cityCurrent) => ({
    type: ActionType.LOAD_PLACES_DATA,
    payload: {placesAll, citiesList, cityCurrent}
  }),

  addFavoritePlace: (favoritePlace) => ({
    type: ActionType.ADD_FAVORITE_PLACE,
    payload: favoritePlace
  })
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.CHANGE_CITY_CURRENT:
      return Object.assign({}, state, {
        cityCurrent: action.payload
      });

    case ActionType.LOAD_PLACES_DATA:
      return Object.assign({}, state, {
        placesAll: action.payload.placesAll,
        citiesList: action.payload.citiesList,
        cityCurrent: action.payload.cityCurrent,
      });

    case ActionType.ADD_FAVORITE_PLACE:
      return Object.assign({}, state, {
        placesAll: state.placesAll.map((item) => {
          if (item.id === action.payload.id) {
            item.isFavorite = action.payload.isFavorite;
          }
          return item;
        })
      });
  }

  return state;
};

const Operations = {
  loadData: () => (dispatch, _, api) => {
    return api.get(REQUEST_URL.HOTELS)
      .then((response) => {
        if (response.status === 200) {
          const placesAll = preparePlacesData(response.data);
          const citiesList = getCitiesList(placesAll);
          const cityCurrent = getRandomCity(citiesList);

          dispatch(ActionsCreator.loadData(placesAll, citiesList, cityCurrent));
        }
      });
  },

  sendFavoriteData: (id, status) => (dispatch, _, api) => {
    return api.post(`${REQUEST_URL.FAVORITE}/${id}/${status}`)
      .then((response) => {
        if (response.status === 200) {
          dispatch(ActionsCreator.addFavoritePlace(preparePlace(response.data)));
        }
      });
  },

  loadFavoritesData: () => (dispatch, _, api) => {
    return api.get(REQUEST_URL.FAVORITE)
      .then((response) => {
        if (response.status === 200) {
          console.log(preparePlacesData(response.data));
        }
      });
  },
};

export {
  Operations,
  ActionsCreator,
  reducer
};
