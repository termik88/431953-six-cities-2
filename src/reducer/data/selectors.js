import {preparePlaces, getCitiesList, getRandomCity, getPlacesSelected} from "../../until.js";
import {createSelector} from "reselect";
import NameSpace from "../name-spaces.js";

const NAME_SPACE = NameSpace.DATA;

export const getPlacesAllSelector = (state) => {
  return preparePlaces(state[NAME_SPACE].placesAll);
};

export const getCitiesListSelector = createSelector(
    getPlacesAllSelector,
    (placesAll) => getCitiesList(placesAll)
);

const getCityCurrent = (state) => {
  return state[NAME_SPACE].cityCurrent;
};

export const getCityCurrentSelector = createSelector(
    getCityCurrent,
    getCitiesListSelector,
    (resultOne, resultTwo) => resultOne ? resultOne : getRandomCity(resultTwo)
);

export const getPlacesSelectedSelector = createSelector(
    getCityCurrentSelector,
    getPlacesAllSelector,
    (resultOne, resultTwo) => getPlacesSelected(resultOne, resultTwo)
);
