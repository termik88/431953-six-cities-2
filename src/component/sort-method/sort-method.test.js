import React from 'react';
import renderer from 'react-test-renderer';

import SortMethod from "./sort-method.jsx";
import {sortMethodsList} from "../../reducer/data/data.js";

describe(`SortMethod component test.`, () => {
  it(`Correctly renders after relaunch`, () => {
    const mockDate = {
      i: 1,
      sortMethodsList,
      sortMethodCurrent: sortMethodsList[0],
      method: sortMethodsList[0],
      onChangeSortMethod: jest.fn(),
      handleToggle: jest.fn()
    };
    const component = renderer
      .create(<SortMethod
        {...mockDate}
      />)
      .toJSON();

    expect(component).toMatchSnapshot();
  });
});
