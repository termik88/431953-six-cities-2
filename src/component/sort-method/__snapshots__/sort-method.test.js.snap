// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SortMethod component test. Correctly renders after relaunch 1`] = `
<li
  className="places__option places__option--active"
  onClick={[Function]}
  tabIndex={1}
>
  Popular
</li>
`;
